src/lib.rs

```diff
@@ -208,75 +187,52 @@ pub trait Decodable {
 /// assert_eq!(decode("zCn8eVZg").unwrap(),
 ///            (Base::Base58btc, b"hello".to_vec()));
 /// ```
-pub fn decode<T: Decodable>(data: T) -> Result<(Base, Vec<u8>), Error> {
+    #[inline]
+pub fn decode<T: Decodable>(data: T) -> Result<(Base, Vec<u8>)> {
     data.decode()
 }
 
-impl<'a> Decodable for &'a str {
-    fn decode(&self) -> Result<(Base, Vec<u8>), Error> {
-        let (base_char, content) = self.split_at(1);
-        let base = try!(Base::from_code(base_char));
+impl Decodable for str {
+    fn decode(&self) -> Result<(Base, Vec<u8>)> {
+        if !self.is_ascii() {
+            return Err(Error::InvalidBaseString);
+        }
+        let code = self.bytes().next().unwrap_or(0);
+        let base = Base::from_code(code)?;
+        let content = &self[1..];
         let alphabet = try!(base.alphabet());
-        let decoded = try!(base_x::decode(&alphabet, content));
-        let res = decoded.iter()
-            .map(|u| *u as u8)
-            .collect();
-        Ok((base, res))
+        let decoded = try!(base::decode(&alphabet, content));
```

Comment:
Could we move all try! to ? to keep it consistent or is there a specific difference?